The errors I'm seeing in CI look to be due to this patch depending on [this pull request against esp-wifi-sys](https://github.com/esp-rs/esp-wifi-sys/pull/504)
```
  error[E0369]: binary operation `==` cannot be applied to type `esp_wifi_sys_esp32c5::include::wifi_sta_config_t`
  Error:     --> /Users/espressif/actions-runner/_work/esp-hal/esp-hal/esp-radio/src/wifi/mod.rs:3263:32
       |
  3263 |                 && current.sta == cfg.sta
       |                    ----------- ^^ ------- esp_wifi_sys_esp32c5::include::wifi_sta_config_t
       |                    |
       |                    esp_wifi_sys_esp32c5::include::wifi_sta_config_t
       |
  note: `esp_wifi_sys_esp32c5::include::wifi_sta_config_t` does not implement `PartialEq`
```
Should I wait for the other PR to get merged (so hopefully the CI checks can clear up) or should I rework this PR so that it doesn't need the other change? I'm happy to go either way.

As for the constant, I'm using at `CONFIG_ESP_WIFI_ESPNOW_MAX_ENCRYPT_NUM` in https://github.com/esp-rs/esp-hal/pull/5849/changes#diff-f50e43beb7d68e871bb177ac89d2a3dfeb6f6b2ead76aaa2e266073760ec77d2R3161 but what I've called `TOTAL_HW_ENCRYPT_KEYS` I can't find an equivalent available constant. The closest, as you pointed out, is `ESP_WIFI_MAX_CONN_NUM` but that's 15 rather than 17 on most devices and as far as I can tell has a different purpose.  It looks to me like it's the size of the station list array in `wifi_sta_list_t`.

--- a/esp-radio/src/wifi/ap.rs
+++ b/esp-radio/src/wifi/ap.rs
@@ -53,6 +53,9 @@ pub struct AccessPointConfig {
     #[builder_lite(reference)]
     pub(crate) password: String,
     /// The maximum number of connections allowed on the access point.
+    /// When set, this number can be clipped to a true upper limit because
+    /// ESPNow and access point connections share a common pool of hardware
+    /// encryption keys.
     #[builder_lite(unstable)]
     pub(crate) max_connections: u16,
     /// Dtim period of the access point (Range: 1 ~ 10).

--- a/esp-radio/src/wifi/mod.rs
+++ b/esp-radio/src/wifi/mod.rs
@@ -3225,7 +3225,8 @@ ignored."
             // Compare the new ap config with the current. Only update if something is changing.
             // This avoids unnecessary connection issues.
             let mut current: wifi_config_t = core::mem::zeroed();
-            if esp_wifi_get_config(wifi_interface_t_WIFI_IF_AP, &mut current) == include::ESP_OK as i32
+            if esp_wifi_get_config(wifi_interface_t_WIFI_IF_AP, &mut current)
+                == include::ESP_OK as i32
                 && current.ap == cfg.ap
             {
                 return Ok(());
@@ -3278,7 +3279,8 @@ ignored."
             // Compare the new sta config with the current. Only update if something is changing.
             // This avoids unnecessary connection issues.
             let mut current: wifi_config_t = core::mem::zeroed();
-            if esp_wifi_get_config(wifi_interface_t_WIFI_IF_STA, &mut current) == include::ESP_OK as i32
+            if esp_wifi_get_config(wifi_interface_t_WIFI_IF_STA, &mut current)
+                == include::ESP_OK as i32
                 && current.sta == cfg.sta
             {
                 return Ok(());

--- a/esp-radio/src/wifi/mod.rs
+++ b/esp-radio/src/wifi/mod.rs
@@ -3168,6 +3168,19 @@ ignored."
         Err(WifiError::Failed)
     }
 
+    // The total hardware encryption key slots available that are shared between
+    // ESP-NOW encrypted peers and the AP connections.
+    // See https://github.com/espressif/esp-idf/blob/master/components/esp_wifi/Kconfig#L589
+    #[cfg(esp32c2)]
+    const TOTAL_HW_ENCRYPT_KEYS: u8 = 4;
+    #[cfg(not(esp32c2))]
+    const TOTAL_HW_ENCRYPT_KEYS: u8 = 17;
+
+    // The upper limit of connections available for the AP. The user set limit in apply_ap_config
+    // is clipped to this value
+    const AP_MAX_CONNECTIONS: u8 =
+        Self::TOTAL_HW_ENCRYPT_KEYS.saturating_sub(CONFIG_ESP_WIFI_ESPNOW_MAX_ENCRYPT_NUM as u8);
+
     fn apply_ap_config(&mut self, config: &AccessPointConfig) -> Result<(), WifiError> {
         let mut cfg = wifi_config_t {
             ap: wifi_ap_config_t {
@@ -3177,7 +3190,9 @@ ignored."
                 channel: config.channel,
                 authmode: config.auth_method.to_raw(),
                 ssid_hidden: if config.ssid_hidden { 1 } else { 0 },
-                max_connection: config.max_connections as u8,
+                // Clip max_connection in the same way as done internally in esp_wifi_set_config.
+                // Doing this here so that we can do easy comparisons below
+                max_connection: (config.max_connections as u8).min(Self::AP_MAX_CONNECTIONS),
                 beacon_interval: 100,
                 pairwise_cipher: wifi_cipher_type_t_WIFI_CIPHER_TYPE_CCMP,
                 ftm_responder: false,
@@ -3206,6 +3221,15 @@ ignored."
             cfg.ap.ssid[0..(config.ssid.len())].copy_from_slice(config.ssid.as_bytes());
             cfg.ap.ssid_len = config.ssid.len() as u8;
             cfg.ap.password[0..(config.password.len())].copy_from_slice(config.password.as_bytes());
+
+            // Compare the new ap config with the current. Only update if something is changing.
+            // This avoids unnecessary connection issues.
+            let mut current: wifi_config_t = core::mem::zeroed();
+            if esp_wifi_get_config(wifi_interface_t_WIFI_IF_AP, &mut current) == include::ESP_OK as i32
+                && current.ap == cfg.ap
+            {
+                return Ok(());
+            }
 
             esp_wifi_result!(esp_wifi_set_config(wifi_interface_t_WIFI_IF_AP, &mut cfg))
         }
@@ -3250,6 +3274,15 @@ ignored."
             cfg.sta.ssid[0..(config.ssid.len())].copy_from_slice(config.ssid.as_bytes());
             cfg.sta.password[0..(config.password.len())]
                 .copy_from_slice(config.password.as_bytes());
+
+            // Compare the new sta config with the current. Only update if something is changing.
+            // This avoids unnecessary connection issues.
+            let mut current: wifi_config_t = core::mem::zeroed();
+            if esp_wifi_get_config(wifi_interface_t_WIFI_IF_STA, &mut current) == include::ESP_OK as i32
+                && current.sta == cfg.sta
+            {
+                return Ok(());
+            }
 
             esp_wifi_result!(esp_wifi_set_config(wifi_interface_t_WIFI_IF_STA, &mut cfg))
         }

